### PR TITLE
Fix VisMF::AsyncWrite FabOnDisk offsets

### LINKED ISSUE #5700: VisMF::AsyncWriteDoit writes wrong FabOnDisk offsets with fab.format!=NATIVE or an empty rank
**Severity:** high · **Category:** correctness · **Subsystem:** Base-io (blast radius 3/5)

**Locations:** `Src/Base/AMReX_VisMF.cpp:2397`, `Src/Base/AMReX_VisMF.cpp:2503`

**Based on commit `248fcfb7a5`** (branch `ai_audit`; line numbers refer to that tree).

VisMF::AsyncWriteDoit builds the _H offset table on the I/O rank from a byte-count model that is supposed to mirror what the background thread writes into the _D_ files. Two places in that model disagree with the writer: the header size is computed with a different FABio than the one used to write, and the per-rank total_bytes gathered by MPI_Gatherv is not consumed for ranks that own no FABs. Either one shifts every later offset and makes VisMF::Read abort.

## The defect

**`Src/Base/AMReX_VisMF.cpp:2397`** — AsyncWriteDoit computes per-FAB byte offsets with the header written by FArrayBox::getFABio() (honours fab.format) but the background thread writes headers with a native FABio_binary (lines 2471/2568); when fab.format is IEEE32 or NATIVE_32 the RealDescriptor strings differ in length, so all recorded offsets are wrong. _(also reported as F014)_

**`Src/Base/AMReX_VisMF.cpp:2503`** — AsyncWriteDoit's header-building loop skips ranks that own no FABs without consuming the one int64 (total_bytes) that MPI_Gatherv placed in globaldata for that rank (rcnt[rank]=1), so every later rank's data is parsed 8 bytes off: wrong FabOnDisk offsets, garbage min/max, and nbytes_on_rank stays -1 which also corrupts offset[] at line 2544.

## Why it matters

*F013:* Any build, runtime fab.format=IEEE32 (or NATIVE_32) plus amrex.async_out=1; VisMF::AsyncWrite(mf,"mf") then VisMF::Read(mf2,"mf"). Reproduced with 1 and 4 ranks: header records offsets 16455/32911/... while the native FAB headers actually occupy 16466 bytes; Read aborts with "FABio::read_header(): expected 'F'". Synchronous VisMF::Write with the same fab.format reads back fine.

*F015:* MPI build, amrex.async_out=1, nprocs>1, any FabArray whose DistributionMapping leaves a rank empty (e.g. 4 ranks, 4 boxes, dm={0,0,3,3}; or a coarse AMR level with fewer boxes than ranks). VisMF::AsyncWrite then VisMF::Read: reproduced -> header shows `FabOnDisk: mf_D_00001 -1` and `... 4656555029671159266`, min/max 1.6e-319, and Read aborts with "FABio::read_header(): expected 'F'" (both async_out_nfiles=2 and =4).

## How to reach it

- *F013:* config: Any build (serial or AMReX_MPI=ON), any SPACEDIM; runtime fab.format=NATIVE_32|IEEE32 (documented in AMReX_FArrayBox.H:214) plus amrex.async_out=1 (Docs IO.rst). Both keys are runtime-only, so every CI config compiles the path.
- *F013:* API path: ParmParse: fab.format=NATIVE_32, amrex.async_out=1 MultiFab mf(ba,dm,nc,0); VisMF::AsyncWrite(mf,"mf"); AsyncOut::Finish(); MultiFab mf2; VisMF::Read(mf2,"mf"); -> second FAB in each _D_ file is read 11 bytes early; FABio::read_header aborts "expected 'F'".
- *F013:* test coverage: none asserting; Tests/AsyncOut/multifab calls VisMF::AsyncWrite with default fab.format and never reads back. Src callers: WriteMultiLevelPlotfile (PlotFileUtil.cpp:222) and AmrLevel::writePlotFile (AmrLevel.cpp:363).
- *F013:* downstream use: ERF Source/IO/ERF_Plotfile.cpp:2173,2176; REMORA Source/IO/REMORA_Plotfile.cpp:700-730; WarpX Source/BoundaryConditions/PML_RZ.cpp:166-169; ERF Exec/RegTests/SDM_Congestus3D*/inputs_* set amrex.async_out=1. incflo src/utilities/io.cpp:713 and WarpX FlushFormatPlotfile.cpp:103 reach it via WriteMultiLevelPlotfile. No app sets fab.format (only user inputs would).
- *F015:* config: AMReX_MPI=ON (path is inside #ifdef BL_USE_MPI, nprocs>1), any SPACEDIM/precision; MPI builds in clang.yml/hypre.yml/petsc.yml/apps.yml. Runtime amrex.async_out=1 (requires MPI_THREAD_MULTIPLE when async_out_nfiles<nprocs).
- *F015:* API path: ParmParse: amrex.async_out=1 // 4 ranks, 4 boxes, DistributionMapping dm(Vector<int>{0,0,3,3}); or any AMR level with fewer boxes than ranks (RoundRobinDoIt with LeastUsedCPUs ordering leaves middle ranks empty) MultiFab mf(ba,dm,1,0); VisMF::AsyncWrite(mf,"mf"); AsyncOut::Finish(); MultiFab mf2; VisMF::Read(mf2,"mf"); -> header offsets -1/garbage, FABio::read_header abort.
- *F015:* test coverage: none; Tests/AsyncOut/multifab uses n_boxes_per_rank=8 (no empty rank) and never reads back. Src callers: WriteMultiLevelPlotfile (PlotFileUtil.cpp:222), AmrLevel::writePlotFile (AmrLevel.cpp:363).
- *F015:* downstream use: ERF Source/IO/ERF_Plotfile.cpp:2173,2176 and Exec/RegTests/SDM_Congestus3D*/inputs_* (amrex.async_out=1); REMORA Source/IO/REMORA_Plotfile.cpp:700-730; WarpX Source/BoundaryConditions/PML_RZ.cpp:166-169; incflo io.cpp:713 and WarpX FlushFormatPlotfile.cpp:103 via WriteMultiLevelPlotfile.

## Suggested fix

Both fixes are local to `AsyncWriteDoit` and independent; land them together since they corrupt the same table and the same read-back test exercises both. **Header size (F013/F014).** The size loop must use the same `FABio` object the worker lambda writes with. Move the `FABio_binary(FPC::NativeRealDescriptor().clone())` construction above the loop and call `fabio->write_header(hss, ...)` there; drop the `FArrayBox::getFABio()` reference. This is consistent with the rest of the function, which already sizes the data with `whichRD = FPC::NativeRealDescriptor()` (line 2361/2399) rather than `FArrayBox::getDataDescriptor()` as the synchronous `Write` path does (line 991). So async output is, and stays, native-only; the fix just makes the header estimate agree with that. **Maintainer decision:** if `fab.format` should instead be honoured by `AsyncWrite` (as `VisMF::Write` does), the worker and the size loop would both need `getFABio()`/`getDataDescriptor()` plus the conversion path; that is a larger change and probably a separate PR. Either way a short note in `IO.rst` saying async output ignores `fab.format` would be worthwhile. **Empty ranks (F015).** The Gatherv layout is `rcnt[rank] = 1 + n_fab_nums*nfabs`, so every rank contributes one leading int64 whether or not it owns FABs. The parse must consume that int64 as it advances `rank`, not only on the first FAB of a rank. The diff moves the `nbytes_on_rank` read into the rank-walking loop so empty ranks take their zero and `offset[]` is correct for the ranks that follow. Trailing empty ranks keep `-1`, which only feeds `offset[]` of ranks with no FABs and is never used. Please add a read-back check to `Tests/AsyncOut/multifab` (a DM with an empty middle rank, and a run with `fab.format=NATIVE_32`), since the existing test never calls `VisMF::Read`.

For `Src/Base/AMReX_VisMF.cpp:2397` (F013):

```diff
--- a/Src/Base/AMReX_VisMF.cpp
+++ b/Src/Base/AMReX_VisMF.cpp
@@ -2379,10 +2379,11 @@
 
     bool strip_ghost = valid_cells_only && mf.nGrowVect() != 0;
 
+    std::shared_ptr<FABio> fabio(new FABio_binary(FPC::NativeRealDescriptor().clone()));
+
     int64_t total_bytes = 0;
     if (localdata.size() > 1) {
         char* pld = (char*)(&(localdata[1]));
-        const FABio& fio = FArrayBox::getFABio();
         for (MFIter mfi(mf); mfi.isValid(); ++mfi)
         {
             std::memcpy(pld, &total_bytes, sizeof(int64_t));
@@ -2394,7 +2395,7 @@
             std::stringstream hss;
             FArrayBox valid_fab(bx, ncomp, false);
             FArrayBox const& header_fab = (strip_ghost) ? valid_fab : fab;
-            fio.write_header(hss, header_fab, ncomp);
+            fabio->write_header(hss, header_fab, ncomp);
             total_bytes += static_cast<std::streamoff>(hss.tellp());
             total_bytes += header_fab.size() * whichRD.numBytes();
 
@@ -2468,8 +2469,6 @@
         }
     }
 
-    std::shared_ptr<FABio> fabio(new FABio_binary(FPC::NativeRealDescriptor().clone()));
-
     AsyncOut::Submit([=] ()
     {
         if (myproc == io_proc)
```

<sub>Sizes each FAB header with the same native FABio_binary the worker thread writes with, so recorded offsets match the file. Data bytes were already sized with the native whichRD, so this makes header and data consistent. AsyncWrite thus always writes native format and ignores fab.format; if maintainers want it honoured, the worker would instead need getFABio()/getDataDescriptor(). Covers F014.</sub>

For `Src/Base/AMReX_VisMF.cpp:2503` (F015):

```diff
--- a/Src/Base/AMReX_VisMF.cpp
+++ b/Src/Base/AMReX_VisMF.cpp
@@ -2496,6 +2496,10 @@
                 {
                     int k = -1;
                     do {
+                        if (nbytes_on_rank[rank] < 0) { // First time for this rank
+                            std::memcpy(&(nbytes_on_rank[rank]), pgd, sizeof(int64_t));
+                            pgd += sizeof(int64_t);
+                        }
                         if (lidx < std::ssize(gidx[rank])) {
                             k = gidx[rank][lidx];
                             ++lidx;
@@ -2508,11 +2512,6 @@
                     hdr->m_min[k].resize(ncomp);
                     hdr->m_max[k].resize(ncomp);
 
-                    if (nbytes_on_rank[rank] < 0) { // First time for this rank
-                        std::memcpy(&(nbytes_on_rank[rank]), pgd, sizeof(int64_t));
-                        pgd += sizeof(int64_t);
-                    }
-
                     int64_t nbytes;
                     std::memcpy(&nbytes, pgd, sizeof(int64_t));
                     pgd += sizeof(int64_t);
```

<sub>Moves the per-rank total_bytes read into the rank-walking do/while so it is consumed before checking whether the rank owns FABs; empty ranks then take their int64 (0) and nbytes_on_rank/offset[] are right. Matches the Gatherv layout (rcnt[rank]=1+...). Ranks after the last non-empty one keep -1, which only feeds offset[] of ranks with no FABs (unused).</sub>

<sub>Diff(s) are against `248fcfb7a5`, written from the current source and verified only with `git apply --check` — never compiled, never run, never applied to the tree. Treat them as precise intent, not tested patches.</sub>

## Verification evidence

<details><summary><code>F013</code> — confirmed (two independent verifier lenses); duplicates folded in: F014</summary>

Lens 1 (refutation attempt): AMReX_VisMF.cpp:2385-2397 computes total_bytes with `const FABio& fio = FArrayBox::getFABio(); fio.write_header(hss, header_fab, ncomp)` (global fab.format), but line 2471 `std::shared_ptr<FABio> fabio(new FABio_binary(FPC::NativeRealDescriptor().clone()))` and 2568 `fabio->write_header(ofs, fab, ...)` write native headers. FABio_binary::write_header (AMReX_FArrayBox.cpp:908-915) emits `"FAB " << *realDesc` so the header length depends on the RealDescriptor (64-bit vs 32-bit differ by 11 chars). m_fod[k].m_head for every FAB after the first in a file is therefore off. Same defect as F014. No dev-branch fix (git log ai_audit..development -- AMReX_VisMF.cpp empty), no open issue.

Lens 2 (reachability/intent): Reproduced with my own scratchpad build (Release, AMReX_MPI=ON, 3D) at 248fcfb7a5: fab.format=NATIVE_32 or IEEE32 + amrex.async_out=1, 1 and 4 ranks, VisMF::AsyncWrite then VisMF::Read aborts 'amrex::Error::N::FABio::read_header(): expected F' (backtrace VisMF::Read->readFAB->read_header). mf_H offsets 0/65614/131229 vs actual native FAB header 89 bytes (descriptor 54 vs 43 chars) => 0/65625/131250. Sync Write with same format reads back exactly; async NATIVE reads back exactly. Cause: AMReX_VisMF.cpp:2385-2399 sizes headers with FArrayBox::getFABio() but 2471/2568 write FABio_binary(NativeRealDescriptor). Introduced 1c24f8a94c (#845); #2564 only reindented. No comment/PR text says async ignores fab.format; git log ai_audit..development empty for this file.

</details>

<details><summary><code>F015</code> — confirmed (two independent verifier lenses)</summary>

Lens 1 (refutation attempt): Gatherv layout (lines 2425-2442): rcnt[rank]=1+n_fab_nums*nfabs, so an empty rank contributes exactly one int64 (localdata[0]=total_bytes=0). Header loop 2494-2504: `do { if (lidx < ssize(gidx[rank])) {...} else { ++rank; lidx = 0; } } while (k < 0);` advances past an empty rank without moving pgd; the next non-empty rank then reads the empty rank's 0 as its nbytes_on_rank (line 2511-2513) and all subsequent int64/Real fields are shifted 8 bytes. nbytes_on_rank[empty] stays -1 and enters `offset[ip] = offset[ip-1] + nbytes_on_rank[ip-1]` (line 2544) when the empty rank is not the last spot in its file, yielding m_head=-1 exactly as reported. Unchanged since 1c24f8a94c (Async out #845); no dev fix, no open issue.

Lens 2 (reachability/intent): Reproduced with my own build, 4 ranks, 8 boxes, amrex.async_out=1: pmap {0,0,0,0,3,3,3,3} (ranks 1,2 empty) gives mf_H 'FabOnDisk: mf_D_00003 -4534276637440606208' (nfiles=4) and 'mf_D_00001 -1' (nfiles=2) and rank 3 aborts in VisMF::Read->readFAB 'expected F'; pmap {1,1,1,1,3,3,3,3} (rank 0 empty) also aborts; trailing empty rank {0,0,0,0,1,1,2,2} and no-empty-rank DMs read back exactly; sync Write with the same DM reads back exactly. Mechanism: MPI_Gatherv rcnt[rank]=1+n_fab_nums*nfabs (2428-2434) so an empty rank contributes one int64, but the loop at 2498-2506 does ++rank without pgd+=8. Introduced 1c24f8a94c (#845); fa795322b44 (#5368) only changed size()->ssize. No development fix, no open issue/PR.

</details>

---

<sub>Based on commit `248fcfb7a5`, the tree the audit verified against. From an automated audit of AMReX `Src/`. Audit finding ids: F013, F015, F014. Reviewer unit(s): Base/IO-VisMF-3. Nothing here was compiled or run except where the evidence says so — the failure scenarios are code reasoning, so the reaching configuration above is the cheapest way to confirm or refute it.</sub>

### PR DESCRIPTION
**Pull request by WeiqunZhang**
The offset table was sized with FArrayBox::getFABio(), which honors fab.format, while the writer thread always writes native FAB headers. Size it with the same native FABio, and document that async output ignores fab.format.

The header loop also skipped a rank owning no FABs without consuming the leading int64 that MPI_Gatherv stored for it, shifting every later rank's data.

Fixes #5700.

### CODE REVIEW

**Comment by WeiqunZhang** [comment]
# PR #5823: Fix VisMF::AsyncWrite FabOnDisk offsets

PR: https://github.com/AMReX-Codes/amrex/pull/5823
Reviewed commit: `53848b9f51aad748914c9e4fb90bc17a6abd30fb`
Reviewed at: 2026-09-07 20:33 UTC
Status: Complete
Result: No actionable findings

## Findings

No actionable findings. Offset sizing and the asynchronous writer share the same native FABio instance. Each visited rank's leading byte count is consumed before searching for a FAB, including empty ranks between data-owning ranks. The documentation accurately describes fab.format behavior.

## Validation

Inspected PR description, conversation, inline comments, and the complete diff with `git diff ae079342654b60f95effddd4377cd9ffaf573f5f 53848b9f51aad748914c9e4fb90bc17a6abd30fb`. Traced local metadata layout, gathered rank traversal, accumulated file offsets, and asynchronous writer capture using git show at the reviewed head. No local async I/O/MPI round-trip test was run in this review. Non-native fab.format values, empty ranks, and multi-file output remain runtime coverage limits.

CI at reviewed head: SUCCESS: 75. CI status is separate from the source assessment.

Reviewed by gpt-6-astra.

**Review by asalmgren — APPROVED** [review]
